python/private/coverage_deps.bzl

```diff
@@ -0,0 +1,105 @@
+        "aarch64-apple-darwin",
+    ),
+    (
+        "pypi__coverage_cp38_aarch64-unknown-linux-gnu",
+        "https://files.pythonhosted.org/packages/40/3b/cd68cb278c4966df00158811ec1e357b9a7d132790c240fc65da57e10013/coverage-6.5.0-cp38-cp38-manylinux_2_17_aarch64.manylinux2014_aarch64.whl",
+        "6c4459b3de97b75e3bd6b7d4b7f0db13f17f504f3d13e2a7c623786289dd670e",
+        "aarch64-unknown-linux-gnu",
+    ),
+    (
+        "pypi__coverage_cp38_x86_64-apple-darwin",
+        "https://files.pythonhosted.org/packages/05/63/a789b462075395d34f8152229dccf92b25ca73eac05b3f6cd75fa5017095/coverage-6.5.0-cp38-cp38-macosx_10_9_x86_64.whl",
+        "d900bb429fdfd7f511f868cedd03a6bbb142f3f9118c09b99ef8dc9bf9643c3c",
+        "x86_64-apple-darwin",
+    ),
+    (
+        "pypi__coverage_cp38_x86_64-unknown-linux-gnu",
+        "https://files.pythonhosted.org/packages/bd/a0/e263b115808226fdb2658f1887808c06ac3f1b579ef5dda02309e0d54459/coverage-6.5.0-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl",
+        "6b07130585d54fe8dff3d97b93b0e20290de974dc8177c320aeaf23459219c0b",
+        "x86_64-unknown-linux-gnu",
+    ),
+    (
+        "pypi__coverage_cp39_aarch64-apple-darwin",
+        "https://files.pythonhosted.org/packages/63/e9/f23e8664ec4032d7802a1cf920853196bcbdce7b56408e3efe1b2da08f3c/coverage-6.5.0-cp39-cp39-macosx_11_0_arm64.whl",
+        "95203854f974e07af96358c0b261f1048d8e1083f2de9b1c565e1be4a3a48cfc",
+        "aarch64-apple-darwin",
+    ),
+    (
+        "pypi__coverage_cp39_aarch64-unknown-linux-gnu",
+        "https://files.pythonhosted.org/packages/18/95/27f80dcd8273171b781a19d109aeaed7f13d78ef6d1e2f7134a5826fd1b4/coverage-6.5.0-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl",
+        "b9023e237f4c02ff739581ef35969c3739445fb059b060ca51771e69101efffe",
+        "aarch64-unknown-linux-gnu",
+    ),
+    (
+        "pypi__coverage_cp39_x86_64-apple-darwin",
+        "https://files.pythonhosted.org/packages/ea/52/c08080405329326a7ff16c0dfdb4feefaa8edd7446413df67386fe1bbfe0/coverage-6.5.0-cp39-cp39-macosx_10_9_x86_64.whl",
+        "633713d70ad6bfc49b34ead4060531658dc6dfc9b3eb7d8a716d5873377ab745",
+        "x86_64-apple-darwin",
+    ),
+    (
+        "pypi__coverage_cp39_x86_64-unknown-linux-gnu",
+        "https://files.pythonhosted.org/packages/6b/f2/919f0fdc93d3991ca074894402074d847be8ac1e1d78e7e9e1c371b69a6f/coverage-6.5.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl",
+        "8f830ed581b45b82451a40faabb89c84e1a998124ee4212d440e9c6cf70083e5",
+        "x86_64-unknown-linux-gnu",
+    ),
+]
+#END: managed by update_coverage_deps.py script
+
+def install_coverage_deps():
+    """Register the dependency for the coverage dep.
+    """
+    for name, url, sha256, platform in _coverage_deps:
+        maybe(
+            http_archive,
+            name = name,
+            build_file_content = """
+py_library(
```

Comment:
I'm not sure this can be a `py_library`.  Doesn't that lead to a dependency when you add it as a dependency for the `python_runtime`?  Or is the python toolchain not actually a dependency of `py_library` targets (as opposed to `py_binary`)?

Reply:
`coverage_tool` needs to be a single file from the toolchain point of view and that is why there is only a single `src` in there. I was following official docs from https://bazel.build/configure/coverage#python and since the manual tests with `bazel coverage` worked in some cases, I thought that was fine?

Reply:
TIL, thanks for explaining. Should we have an extra line of docs somewhere?

Reply:
I would suggest not using py_library for now -- it implies a set of functionality that simply isn't respected (e.g. imports attribute won't be respected, anything in its PyInfo is ignored, etc).

It's currently possible to put a py_library there, but is a bit risky. This is because py_library doesn't, today, depend on the toolchain. However, I don't see how it can avoid that forever (in particular, I'm thinking of for when pyc precompilation is added).

I _do_ think a py_library-like-equivalent dependency is the correct conceptual type of dependency that should be there, though. Running with code coverage is, essentially, adding an extra dependency to the binary and wrapping the regular entry point in another entry point. Unfortunately, it's not immediately clear to me how to do make that work cleanly via (or with support from) the toolchain (I have a few ideas, but haven't thought them through).

To use the coverage settings of the toolchain, you need to:
1. Point coverage_tool to a single file. Only this file is consumed. Any additional files, runfiles, dependencies, etc aren't used.
2. Put any files coverage_tool needs in coverage_files, which will be added to the binary's runfiles.

Reply:
I mean `filegroup` is [clearly documented](https://bazel.build/reference/be/general#filegroup.srcs) that `srcs` is in the files output for the rule (which we need) and `data` is in the `runfiles`.  I don't think we need to make it more complicated than that (it is also true that bazel will treat such a `filegroup` as an executable target, however we don't think we actually depend on that behavior).

Reply:
Er, sorry, I mixed up the py_runtime and PyRuntimeInfo apis. Disregard the "you have to specify two attributes" part. The splitting of the target into (entry point, files) is handled by py_runtime.

Reply:
> > I don't think we need to make it more complicated than that
> 
> I think you're referring to my "it should be a py library-like-equivalent dependency" comment?
> 
> There's a big distinction between "bundle of pre-built, platform-specific files" and a "dependency" (for lack of a better term). The former carries no information other than "here's a bag of files". A dependency can carry additional information, such as modifications to make to the import path, additional linker flags, propagating a _proper_ runfiles object (we currently omit symlinks, for example), or other necessary metadata.

Yes, and in this particular case, so long as the target has its visibility restricted to the toolchain, a bag of pre-built platform-specific files is exactly what this is - the `py_runtime` rule won't pay any attention to additional providers from the target, and if it did honor e.g. `imports` from this input, I can't see a situation where that wouldn't do more harm than good.  But, this is a very good reason to make sure the visibility is appropriately restricted.

> "depend" is a strong word, but we do allow the "executable" of the target to be an indicator as to what the entry point is. The basic logic for converting the Target to the (entry point, files) is:
> 
> * If there is one output, use that
> * otherwise, use whatever the executable file is
> * Otherwise, error

Yes, I do remember how I implemented that.  But in this case I was referring to the fact that a `filegroup` with a single file in `src` will have a `FilesToRun` provider supplying that file as an executable, which would make the second option you list above there work, but we don't depend on that behavior (which is, admittedly, a little weird) because it'll stop at the first in this case.

At any rate, if a `filegroup` is what we want to use here, then the upstream docs in the bazel repo should maybe be updated?  Except that there _might_ be use cases for having it as a `py_library` in the general case; just in this particular case we don't need it to be and by using restricted visibility and a `filegroup` we can afford to not think about all of the ways that could go wrong.

Reply:
I changed it to use filegroup, so it should be hopefully an improvement over `py_library` and we can figure out something better as a separate PR.